src/vcgen.cpp

```diff
@@ -451,6 +462,84 @@ optional<string> encodeOp(State &st, mlir::memref::BufferCastOp op) {
   return {};
 }
 
+vector<z3::expr> getSizes(const State &st, mlir::memref::SubViewOp op) {
+  vector<z3::expr> sizes;
+  for (unsigned i = 0; i < op.getSourceType().getRank(); i ++) {
+    if (op.isDynamicSize(i)) {
+      sizes.push_back(st.regs.get<Index>(op.getDynamicSize(i)));
+    } else {
+      sizes.push_back(Index(op.getStaticSize(i)));
+    }
+  }
+  return sizes;
+}
+
+vector<z3::expr> getOffsets(const State &st, mlir::memref::SubViewOp op) {
+  vector<z3::expr> offsets;
+  for (unsigned i = 0; i < op.getSourceType().getRank(); i ++) {
+    if (op.isDynamicOffset(i)) {
+      offsets.push_back(st.regs.get<Index>(op.getDynamicOffset(i)));
+    } else {
+      offsets.push_back(Index(op.getStaticOffset(i)));
+    }
+  }
+  return offsets;
+}
+
+vector<z3::expr> getStrides(const State &st, mlir::memref::SubViewOp op) {
+  vector<z3::expr> strides;
+  for (unsigned i = 0; i < op.getSourceType().getRank(); i ++) {
+    if (op.isDynamicStride(i)) {
+      strides.push_back(st.regs.get<Index>(op.getDynamicStride(i)));
+    } else {
+      strides.push_back(Index(op.getStaticStride(i)));
+    }
+  }
+  return strides;
+}
+
+template<>
+optional<string> encodeOp(State &st, mlir::memref::SubViewOp op) {
+  auto sizes = getSizes(st, op);
+  auto offsets = getOffsets(st, op);
+  auto strides = getStrides(st, op);
+  auto src = st.regs.get<MemRef>(op.source());
+  auto layout = src.toSubViewLayout(offsets, strides);
+
+  int rankDiff = op.getSourceType().getRank() - op.getType().getRank();
+  assert(rankDiff >= 0); // only reducing rank is allowed
+
+  if (rankDiff > 0) {
+    vector<z3::expr> reducedSizes, reducedOffsets, reducedStrides;
+    vector<z3::expr> reducedIndVars, substIndVars, substConsts;
+    for (unsigned i = 0; i < sizes.size(); i++) {
+      if (rankDiff > 0 && sizes[i].is_numeral() && sizes[i].as_uint64() == 1) { //statically known to be of size 1
+        substIndVars.push_back(layout.indVars[i]);
+        substConsts.push_back(Index::zero());
+        rankDiff --;
+      } else {
+        reducedSizes.push_back(sizes[i]);
+        reducedOffsets.push_back(offsets[i]);
+        reducedStrides.push_back(strides[i]);
+        reducedIndVars.push_back(layout.indVars[i]);
+      }
+    }
+    z3::expr reducedExpr = layout.expr
+      .substitute(toExprVector(substIndVars), toExprVector(substConsts));
+    z3::expr reducedInbounds = layout.inbounds
+      .substitute(toExprVector(substIndVars), toExprVector(substConsts));
+    auto reducedLayout = MemRef::Layout(reducedIndVars, reducedExpr, reducedInbounds);
+    auto memref = MemRef(st.m.get(), src.getBID(), src.getOffset(),
+      reducedSizes, reducedLayout, Float::sort());
+    st.regs.add(op.getResult(), move(memref));
+
+  } else {
+    auto memref = MemRef(st.m.get(), src.getBID(), src.getOffset(),  sizes, layout, Float::sort());
+    st.regs.add(op.getResult(), move(memref));
```

Comment:
Here we pass source memref's bid, offset expression.
So the result memref has same variable of source's which makes aliasing result.